### DISCUSSION TITLE
Some advice for ModbusClientRTU usage

### DISCUSSION THREAD

**Discussion by Miq1** [eModbus/eModbus#76] [General]
I got bitten myself, so I would like to share my experience:

I had a ``ModbusClientRTU`` running on a ``ModbusBridgeWiFi`` device and noticed sporadic CRC errors when requesting the RTU client. A closer examination revealed that the ``serial.read()`` used to get the response was interrupted for a few milliseconds, when the error occurred. 

The reason was that both the bridge and client tasks were running on core 0. During the ``read()``, the bridge task sometimes got a time slot that led to the interrupted reads. I optimized the receive routine on and on without being able to completely prevent this.

So the only remedy and my advice is: *use ``begin(1)`` for the ``ModbusClientRTU``!* This will use core 1 for it where it can work undisturbed.

**Comment by bertmelis** [comment]
You mean the FreeRTOS scheduler messes up the timing restrictions?
Aren't there FreeRTOS macros to avoid this? (Not sure about this, it might disable interrupts and everything)
Or is giving the RTU comm task a high priority helping?

**Reply by Miq1** [reply]
> You mean the FreeRTOS scheduler messes up the timing restrictions?

No, not that I am aware of it.

> Aren't there FreeRTOS macros to avoid this? (Not sure about this, it might disable interrupts and everything)

I did not look for any - may be?

> Or is giving the RTU comm task a high priority helping?

No, that I did try. It looks like the scheduler is yielding to tasks during the ``serial.read()`` itself, on a deeper level. 

As I wrote, it is only a matter of a few milliseconds (less than 4), but in our case it matters, unfortunately. Regular tasks using serial probably will not notice. Even more, as in the regular core the UART FIFO trigger will run only after 112 bytes have been received, as we learned. 4ms is about 7 bytes transmitted at 19200, so no one might ever considered that a problem.
